streetcrm/views.py

```diff
@@ -679,3 +716,77 @@ def get_context_data(self, **context):
         return context
 
 
+class EventParticipantsImport(FileImportMixin, generic.View):
+    field_processors = {
+        "institution": process_institution_name
+    }
+    # Mapping of CSV header row names to model fields
+    field_map = {
+        "Name": "name",
+        "Institution": "institution",
+        "Phone Number": "primary_phone",
+        "Email": "email"
+    }
+    # Django query syntax for each field
+    field_queries = {
+        "name": "name__iexact",
+        "email": "email__iexact"
+    }
+
+    def import_data(self, rows):
+        event = models.Event.objects.get(pk=self.kwargs['pk'])
+        event_participant_ids = event.participants.all().values_list('id', flat=True)
+
+        # Create list of participants to create and add for bulk 
+        # create/update operatiosn
+        participants_to_create = []
+        participants_to_add = []
+        for row in rows:
+            # Use field queries (with default of the key text) to check for
+            # record existence
+            participant_query = models.Participant.objects.filter(
+                **{self.field_queries.get(key, key): value for key, value in row.items()}
+            )
+            # If a participant is found, add them to the event if not already
+            if participant_query:
+                participant = participant_query[0]
+                if participant.id not in event_participant_ids:
+                    participants_to_add.append(participant)
+            # Otherwise add them to the list of participants to create
+            else:
+                participants_to_create.append(models.Participant(**row))
+        
+        # Create all new participants, then add them as well as existing
+        # participants not associated with the event to its participants
+        new_participants = models.Participant.objects.bulk_create(participants_to_create)
+        participant_ids_to_add = (
+            [p.id for p in participants_to_add] +
+            [np.id for np in new_participants]
```

Comment:
Ah, so this is the problematic part, in that we don't actually know the IDs of the newly created participants?